File: src/main/scala/com/abaddon83/cardsGames/burracoGames/commands/EndGameCmd.scala
package com.abaddon83.cardsGames.burracoGames.commands

import java.util.UUID

import com.abaddon83.cardsGames.burracoGames.ports.GameRepositoryPort
import com.abaddon83.cardsGames.shares.games.GameIdentity
import com.abaddon83.cardsGames.shares.players.PlayerIdentity
import com.abaddon83.libs.cqs.commands.{Command, CommandHandler}

import scala.concurrent.{Await, Future}

case class EndGameCmd(
                         gameIdentity: GameIdentity,
                         playerIdentity: PlayerIdentity
                       ) extends Command {
  override protected val requestId: UUID = UUID.randomUUID()
}

class EndGameHandler(
                           gameRepositoryPort: GameRepositoryPort
                         )
                         (implicit val ec: scala.concurrent.ExecutionContext)
  extends CommandHandler[EndGameCmd] {

  override def handleAsync(command: EndGameCmd): Future[Unit] = {
    for {
      gameEnd <- gameRepositoryPort.findBurracoGameInitialisedTurnEndBy(gameIdentity = command.gameIdentity)
      updatedGameEnd = gameEnd.completeGame(playerIdentity = command.playerIdentity)
    } yield gameRepositoryPort.save(updatedGameEnd)
  }

  override def handle(command: EndGameCmd): Unit = {
    import scala.concurrent.duration._
    import scala.language.postfixOps
    val gameEnd = Await.result(gameRepositoryPort.findBurracoGameInitialisedTurnEndBy(gameIdentity = command.gameIdentity), 500 millis)
    val updatedGameEnd = gameEnd.completeGame(playerIdentity = command.playerIdentity)
    gameRepositoryPort.save(updatedGameEnd)
  }

}

File: src/main/kotlin/com/abaddon83/burracoGame/commands/EndGameCmd.kt
package com.abaddon83.burracoGame.commands

import com.abaddon83.burracoGame.domainModels.PlayerNotAssigned
import com.abaddon83.burracoGame.ports.BurracoGameRepositoryPort
import com.abaddon83.burracoGame.shared.games.GameIdentity
import com.abaddon83.burracoGame.shared.players.PlayerIdentity
import com.abaddon83.utils.cqs.commands.Command
import com.abaddon83.utils.cqs.commands.CommandHandler
import kotlinx.coroutines.runBlocking
import org.koin.core.KoinComponent
import org.koin.core.inject
import java.sql.Timestamp
import java.util.*

data class EndGameCmd(
        val gameIdentity: GameIdentity,
        val playerIdentity: PlayerIdentity,
        override val commandId: UUID,
        override val commandTimeStamp: Timestamp): Command {

    constructor(gameIdentity: GameIdentity,
                playerIdentity: PlayerIdentity): this(
            gameIdentity = gameIdentity,
            playerIdentity = playerIdentity,
            commandId = UUID.randomUUID(),
            commandTimeStamp = Timestamp(System.currentTimeMillis())
    )
}


class EndGameHandler() : CommandHandler<EndGameCmd>, KoinComponent {

    private val repository: BurracoGameRepositoryPort by inject()

    override fun handle(command: EndGameCmd) {
        val gameEnd = runBlocking { repository.findBurracoGameExecutionTurnEndBy(gameIdentity = command.gameIdentity) }
        check(gameEnd != null) {"GameIdentity ${command.gameIdentity} not found exist"}
        val updatedGameEnd = gameEnd.completeGame(playerIdentity = command.playerIdentity)
        repository.save(updatedGameEnd)
    }

    override suspend fun handleAsync(command: EndGameCmd) {
        val gameEnd = repository.findBurracoGameExecutionTurnEndBy(gameIdentity = command.gameIdentity)
        check(gameEnd != null) {"GameIdentity ${command.gameIdentity} not found exist"}
        val updatedGameEnd = gameEnd.completeGame(playerIdentity = command.playerIdentity)
        repository.save(updatedGameEnd)
    }
}